common/web/keyboard-processor/src/keyboards/keyboardLoaderBase.ts

```diff
@@ -19,5 +20,48 @@ export default abstract class KeyboardLoaderBase {
     return promise;
   }
 
-  protected abstract loadKeyboardInternal(uri: string): Promise<Keyboard>;
+  public loadKeyboardFromStub(stub: KeyboardProperties & { filename: string }) {
+    this.harness.install();
+    let promise = this.loadKeyboardInternal(stub.filename, stub.id);
+
+    promise = promise.catch((err: Error) => {
+      if(err == KeyboardLoaderBase.scriptErrorMessage(stub.filename)) {
+        // Enhance the error message.
+        err.message = KeyboardLoaderBase.scriptErrorMessage(stub);
+      } else if(err == KeyboardLoaderBase.missingErrorMessage(stub.filename)) {
+        // Same thing here.
+        err.message = KeyboardLoaderBase.scriptErrorMessage(stub);
+      }
+
+      throw err;
+    })
+
+    return promise;
+  }
+
+  protected abstract loadKeyboardInternal(uri: string, id?: string): Promise<Keyboard>;
+
+  protected static scriptErrorMessage(uri: string);
+  protected static scriptErrorMessage(metadata: KeyboardProperties & { filename: string });
+  protected static scriptErrorMessage(arg: string | (KeyboardProperties & { filename: string })) {
+    if(typeof arg == "string") {
+      const uri = arg;
+      return `Error registering the keyboard script at ${uri}; it may contain an error.`
+    } else {
+      const stub = arg;
+      return `Error registering the ${stub.name} keyboard for ${stub.langName}; keyboard script at ${stub.filename} may contain an error.`
+    }
+  }
+
+  protected static missingErrorMessage(uri: string);
+  protected static missingErrorMessage(metadata: KeyboardProperties & { filename: string });
+  protected static missingErrorMessage(arg: string | (KeyboardProperties & { filename: string })) {
+    if(typeof arg == "string") {
+      const uri = arg;
+      return `Cannot find the keyboard at ${uri}.`
+    } else {
+      const stub = arg;
+      return `Cannot find the ${stub.name} keyboard for ${stub.langName} at ${stub.filename}.`;
+    }
+  }
```

Comment:
These feel like a lot of syntactic sugar and typechecking for could essentially be separate error messages:

```
  protected static scriptUriErrorMessage(uri: string) {
    return `Error registering the keyboard script at ${uri}; it may contain an error.`;
  }

  protected static scriptStubErrorMessage(stub: KeyboardProperties & { filename: string }) {
    return `Error registering the ${stub.name} keyboard for ${stub.langName}; keyboard script at ${stub.filename} may contain an error.`;
  }
```

(That second one also feels painful because I don't really understand where filename is coming from in this instance)